File: frontend/lib/ai/tool-source.ts
import type { ToolSource } from "@/lib/ai/types";
import type { ToolCallSource } from "@/store";

/** Convert AI event source to store source (snake_case to camelCase) */
export function convertToolSource(source?: ToolSource): ToolCallSource | undefined {
  if (!source) return undefined;
  if (source.type === "main") return { type: "main" };
  if (source.type === "sub_agent") {
    return {
      type: "sub_agent",
      agentId: source.agent_id,
      agentName: source.agent_name,
    };
  }
  if (source.type === "workflow") {
    return {
      type: "workflow",
      workflowId: source.workflow_id,
      workflowName: source.workflow_name,
      stepName: source.step_name,
      stepIndex: source.step_index,
    };
  }
  return undefined;
}

File: backend/crates/golish-core/src/events/tool_source.rs
//! [`ToolSource`] — origin of a tool call (main agent / sub-agent / workflow).

use serde::{Deserialize, Serialize};

/// Source of a tool call - indicates where the tool request originated.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ToolSource {
    /// Tool called by the main agent
    #[default]
    Main,
    /// Tool called by a sub-agent
    SubAgent {
        agent_id: String,
        agent_name: String,
    },
    /// Tool called by a workflow
    Workflow {
        workflow_id: String,
        workflow_name: String,
        /// Current step name (if within a step)
        #[serde(skip_serializing_if = "Option::is_none")]
        step_name: Option<String>,
        /// Current step index (0-based)
        #[serde(skip_serializing_if = "Option::is_none")]
        step_index: Option<usize>,
    },
}
